### ISSUE TITLE
Resource Gallery "See More" opens incorrect modal

### ISSUE THREAD

**Issue by jukent** [ProjectPythia/projectpythia.github.io#235]
On the resource gallery, when I click the "... More" to open the modal with the full card description, it opens the incorrect modal. I so far only noticed this with the "An Introduction to Earth and Environmental Data Science" card, which opens the Unidata Python Training modal, instead of the expected modal

![image](https://user-images.githubusercontent.com/46687291/166570544-bdb6310f-51c8-4f2d-a65f-bc41483f2e17.png)

![image](https://user-images.githubusercontent.com/46687291/166570593-350cb5e7-0040-46c6-a59f-f8f0c71773fb.png)

I haven't yet gone through the gallery to see how pervasive this is, but I'll comment an update.

**Comment by jukent** [comment]
All of the modals seem to be off by one (display the previous card that had a "... More"'s modal instead of their own).

**Comment by jukent** [comment]
<img width="404" alt="Screen Shot 2022-05-05 at 9 04 34 AM" src="https://user-images.githubusercontent.com/46687291/166953824-89ce5e16-6a31-4b9f-9b7d-a77ce35eae32.png">

This shows that the modal content is being correctly generated, but the wrong modal's display turns from "None" to on when the "see more" button is clicked.

**Comment by kmpaul** [comment]
@jukent:  I've taken a little look into this and I think I can see the problem.  I did a *right-click &rarr; view source* on the gallery page to view the HTML source of the gallery page.  Then, I did a <kbd>CTRL-F</kbd> "find in page" and looked for the strings `class="modal"`, `class="modal-btn"`, and `class="modal-background"`.  Sure enough, I found only 1 `class="modal-background"`, as you would hope.  However, I found 11 `class="modal-btn"` strings and 18 `class="modal"` strings!  

Why is this a problem?  Because the algorithm for triggering the appropriate `class="modal"` object (and changing its `display: none;` to/from `display: block;`) when the corresponding `class="modal-btn"` is clicked is based on the arrays/lists of each kind of object (`class="modal"` vs `class="modal-btn"`) being *the same length*.  That is, when `class="modal-btn"` `i` is clicked, it triggers the display of `class="modal"` `i`.  But if the length of the arrays is different, then there is an incorrect mapping from `class="modal-btn"` element to `class="modal"` element.

So, I think the trick is to look at the gallery generator extension code (https://github.com/ProjectPythia/projectpythia.github.io/blob/main/portal/_extensions/yaml_gallery_generator.py) and see why a `class="modal"` element might be created when a corresponding `class="modal-btn"` element is *not* created.

**Comment by kmpaul** [comment]
And I suspect that the error is somewhere in the `truncate()` function in `truncatehtml.py`:

https://github.com/ProjectPythia/projectpythia.github.io/blob/937134f8030fddbd7a6a893bc14073a517525e3b/portal/_extensions/truncatehtml.py#L146

**Comment by jukent** [comment]
Thanks @kmpaul I was attempting to use an indexed ID to force the mapping between the 2. This gets to the heart of the problem. Thanks for the direction.

**Comment by jukent** [comment]
All of the "Your First Python tutorial" recordings, as well as the Unidata Python Training card, have a modal without a modal button. This seems to be all of the instances of modals =/= modal buttons.

For the "Your First ... " They all end in a link, the entire text is displayed on the card correctly and no modal is needed, yet one is generated. 

On the Unidata card, the text is truncated too short stopping before the "&amp" character, which might be uniquely interacting with the truncate function.

**Comment by kmpaul** [comment]
Yeah.  I think you're on the right track.  I think there are at least 2 problems.

One problem stems from the fact that the condition that the `yaml_gallery_generator.py` code uses to determine whether to create a model is based on whether `len(item['description']) > max_descr_len`.  But the `Tokenizer` in `truncatehtml.py` only adds the `ellipsis` string if it counts the number of *rendered HTML characters* (e.g. `&amp;` counts as 1 character, and `<tag stuff="thing"></tag>` counts as 0 characters) is greater than `max_descr_len`.  So, when you have `item['description']` strings that contain HTML in them, the actual string length might be greater than `max_descr_len` (triggering the creation of the modal) but the *rendered* string length might be less than `max_descr_len` (*not* triggering the creation of the modal button).

The solution I think involves defining special "measure the rendered HTML length of the string" function, like this:

```python
def lenhtml(s):
    length = 0  # number of characters (not counting markup) placed in retval so far
    tokens = th.Tokenizer(s)
    tok = tokens.next_token()
    while tok != -1:
        if not isinstance(tok, (th.OpenTag, th.CloseTag, th.SelfClosingTag)):
            length += 1
        tok = tokens.next_token()
    return length
```

**Comment by kmpaul** [comment]
I haven't figured out why the Unidata card is truncated at the & symbol.  When you use the `truncatehtml.truncate` function as it is used in the `yaml_gallery_generator.py` module on that exact description string, it returns the correct result:

```python
>>> s = 'Introduction to Python for Atmospheric Science &amp; Meteorology. Unidata is working to create a collection of online training materials focused on the use of Python in the atmospheric sciences. While our examples and scenarios may feature Unidata tools and data technologies, our aim is to present a generic set of freely available tools that are generally useful to scientists, educators, and students in the geosciences, broadly defined.'

>>> import truncatehtml as th

>>> th.truncate(s, 300, ellipsis='<a class="modal-btn"> ...more</a>')
'Introduction to Python for Atmospheric Science &amp; Meteorology. Unidata is working to create a collection of online training materials focused on the use of Python in the atmospheric sciences. While our examples and scenarios may feature Unidata tools and data technologies, our aim is to present a generic<a class="modal-btn"> ...more</a>'
```

So, it's not the `truncate` function that is the problem with that one.

**Comment by jukent** [comment]
Hmm, but the truncate function is being called and creating a short description that is super short. I'm trying changing the `next_token` function to recognize "&" as "&amp;" to see what happens.

**Comment by kmpaul** [comment]
Yeah, but I can't figure out why it is creating a short description because it appears to work perfectly fine in a simple test.  So, you should add some debugging output to the generator extension to see if the the truncate function really is working properly.

**Comment by jukent** [comment]
Okay. I tested the _len_html function, it isn't quite right yet - still generating extra modals. I have a meeting for the next 2 hours and won't be able to look at this again till our meeting this afternoon.

**Comment by kmpaul** [comment]
Me too. 

Like I said, I think there may be multiple problems and the `len_html` function only fixes one.

**Comment by kmpaul** [comment]
@jukent:  Ok.  I dug into this and I think I found the best solution to the problem.  Some my comments in your PR #236 suggest how to fix it properly.  At the end of the day, I think the `_len_html` function isn't needed at all.
